catalyst/callbacks/mixup.py

```diff
@@ -1,101 +1,185 @@
+                "criterion": dl.CriterionCallback(
+                    metric_key="loss", input_key="clf_logits", target_key="clf_targets_one_hot",
+                ),
+                "optimizer": dl.OptimizerCallback(metric_key="loss"),
+                "classification": dl.ControlFlowCallback(
+                    dl.PrecisionRecallF1SupportCallback(
+                        input_key="clf_logits", target_key="clf_targets", num_classes=10,
+                    ),
+                    ignore_loaders="train",
+                ),
+            },
+        )
+
+    .. note::
+        With running this callback, many metrics (for example, accuracy) become undefined, so
+        use ControlFlowCallback in order to evaluate model(see example)
+    """
+
+    def __init__(
+        self, keys: Union[str, List[str]], alpha=0.2, mode="replace", on_train_only=True, **kwargs,
+    ):
+        """
+
+        Args:
+            keys: batch keys to which you want to apply augmentation
+            alpha: beta distribution a=b parameters. Must be >=0. The more alpha closer to zero the
+                less effect of the mixup.
+            mode: mode determines the method of use. Must be in ["replace", "add"]. If "replace"
+                then replaces the batch with a mixed one, while the batch size is not changed
+                If "add", concatenates mixed examples to the current ones, the batch size increases
+                by 2 times.
+            on_train_only: apply to train only. As the mixup use the proxy inputs, the targets are
+                also proxy. We are not interested in them, are we? So, if on_train_only is True,
+                use a standard output/metric for validation.
+            **kwargs:
+        """
+        assert isinstance(keys, (str, list)), f"keys must be str of list[str], get: {type(keys)}"
+        assert alpha >= 0, "alpha must be>=0"
+        assert mode in ["add", "replace"], f"mode must be in 'add', 'replace', get: {mode}"
+        super().__init__(order=CallbackOrder.Internal)
+        if isinstance(keys, str):
+            keys = [keys]
+        self.keys = keys
+        self.on_train_only = on_train_only
+        self.alpha = alpha
+        self.mode = mode
+        self.required = True
+
+    def _handle_batch(self, runner: "IRunner") -> None:
+        """
+        Applies mixup augmentation for a batch
+
+        Args:
+            runner: runner for the experiment.
+        """
+        batch_size = runner.batch[self.keys[0]].shape[0]
+        beta = np.random.beta(self.alpha, self.alpha, batch_size).astype(np.float32)
+        indexes = np.array(list(range(batch_size)))
+        # index shift by 1
+        indexes_2 = (indexes + 1) % batch_size
+        for key in self.keys:
+            targets = runner.batch[key]
+            targets_shape = [batch_size] + [1] * len(targets.shape[1:])
+            key_beta = torch.Tensor(beta.reshape(targets_shape))
```

Comment:
do we need `device` here (something like `device=runner.engine.device` maybe?)? 
could you please check the minimal example on colab gpu for correctness?